### PR TITLE
Introduce grpc.Formatter abstraction to support non-protobuf IDL relatively easily.

### PR DESCRIPTION
**Pull request by iamqizhao**
The generated code for interop has not been updated. Will do after we reach the agreement.

### CODE REVIEW

**Comment by iamqizhao** [comment]
Rename Formatter to Marshaler. I do not have good idea regarding RecvMessage and SendMessage though. But I thin their names are probably just fine -- send/recv a message via the input marshaler. Let me know if you have better idea.

File: marshaler.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+package grpc
+
+import (
+	"github.com/golang/protobuf/proto"
+)
+
+// Marshaler is an abstraction over message marshaling and unmarshaling in order
+// to support various message formats.
+type Marshaler interface {
+	marshal() ([]byte, error)
```
Comment (dsymonds) [inline marshaler.go]: By making these unexported, this interface can not be implemented by other packages. I don't think that's what you intended.

File: marshaler.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	unmarshal(buf []byte) error
+}
+
+type protoMessage struct {
+	proto.Message
+}
+
+func (m *protoMessage) marshal() ([]byte, error) {
```
Comment (dsymonds) [inline marshaler.go]: No need to put these on `*protoMessage`. These methods can happily be on `protoMessage` (just drop the ampersand inside the constructor func).

**Comment by iamqizhao** [comment]
close. will have a new PR for this per the discussion with dsymonds, sameer and bcmills.

**Pull request closed without merging**